packages/flutter/lib/src/painting/text_painter.dart

```diff
@@ -426,6 +433,90 @@ class TextPainter {
   }
   List<PlaceholderDimensions>? _placeholderDimensions;
 
+  // Placeholders when used in TextFields are frequently inserted as replacements
+  // for strings of text. However, in text layout, the placeholder is treated as
+  // a single object replacement character codepoint. This means that the TextField
+  // treats text editing positions/offsets with the full un-replaced string's length
+  // while the layout and rendering treats placeholders as a single offset.
+  //
+  // This method converts a un-replaced offset to the layout engine's single codepoint
+  // placeholder offset by accounting for the replaced string's length and subtracting
+  // it from the offset.
+  //
+  // This conversion should be applied before calling any _paragraph methods, and any
+  // offsets returned by _paragraph should call _getRawOffset before using the value.
+  //
+  // See also:
+  //
+  //  * _getRawOffset
+  //  * _getPlaceholderAdjustedPosition
+  int _getPlaceholderAdjustedOffset(int offset) {
+    if (_placeholderDimensions == null) {
+      return offset;
+    }
+    int adjustment = 0;
+    for (final PlaceholderDimensions dims in _placeholderDimensions!) {
+      if (dims.range != null) {
+        if (dims.range!.end <= offset) {
+          // placeholders are represented as a single replacement character,
+          // so we subtract 1 from the length to account for it.
+          adjustment += dims.range!.end - dims.range!.start - 1;
+        } else if (dims.range!.end > offset && offset >= dims.range!.start) {
+          // Within the range.
+          // place offset at beginning or end of placeholder depending on
+          // which half it is in.
+          adjustment += offset - dims.range!.start;
+          if (offset > dims.range!.start + dims.range!.end / 2) {
```

Comment:
I have added affinity support, but for caret positioning from tapping, I have still used the "which half it is in" since the affinity of the tap is ambiguous and we default to always picking one affinity first to try before the other, making it extremely difficult to select one side of the widget. The 'which half' method seems to be much more consistent and desirable UX.

For everything else, I pass the affinity in and we use that to determine positioning.